File: src/main/kotlin/school/charset/app/domain/encoding/CodePoint.kt
package school.charset.app.domain.encoding

/**
 * Unicode code point. The valid range is U+0000..U+10FFFF; values inside the
 * surrogate range U+D800..U+DFFF are still legal `CodePoint`s (they exist in
 * the Unicode standard as reserved values), but they are not valid characters
 * and will be rejected by every encoder - that constraint lives in the encoders.
 */
@JvmInline
value class CodePoint(val value: Int) {
    init {
        require(value in MIN..MAX) { "Code point out of range: 0x${value.toString(16).uppercase()}" }
    }

    val isSurrogate: Boolean
        get() = value in SURROGATE_MIN..SURROGATE_MAX

    val isBmp: Boolean
        get() = value <= BMP_MAX

    override fun toString() = "U+%04X".format(value)

    companion object {
        const val MIN = 0x0000
        const val MAX = 0x10FFFF
        const val SURROGATE_MIN = 0xD800
        const val SURROGATE_MAX = 0xDFFF
        const val BMP_MAX = 0xFFFF // Basic Multilingual Plane
    }
}

File: lib/charset/encoding/code_point.ex
defmodule Charset.Encoding.CodePoint do
  @moduledoc """
  Unicode code point guards and formatting.

  A code point is a plain integer in `0..0x10FFFF` - no wrapper struct, so the
  guards compose directly into function heads (and feed the compiler's type
  inference). Values inside the surrogate range U+D800..U+DFFF are still legal
  code points (they exist in the Unicode standard as reserved values), but
  they are not valid characters and are rejected by every encoder - that
  constraint lives in the codec.
  """

  @min 0x0000
  @max 0x10FFFF
  @surrogate_min 0xD800
  @surrogate_max 0xDFFF
  # Basic Multilingual Plane
  @bmp_max 0xFFFF

  defguard is_code_point(cp) when is_integer(cp) and cp in @min..@max

  defguard is_surrogate(cp) when is_integer(cp) and cp in @surrogate_min..@surrogate_max

  defguard is_bmp(cp) when is_integer(cp) and cp in @min..@bmp_max

  def min, do: @min
  def max, do: @max
  def surrogate_min, do: @surrogate_min
  def surrogate_max, do: @surrogate_max
  def bmp_max, do: @bmp_max

  @doc """
  Formats a code point in the standard `U+XXXX` notation (4 hex digits
  minimum, more when needed).

      iex> Charset.Encoding.CodePoint.format(0xE9)
      "U+00E9"

      iex> Charset.Encoding.CodePoint.format(0x1F389)
      "U+1F389"
  """
  @spec format(integer()) :: String.t()
  def format(cp) when is_code_point(cp) do
    "U+" <> String.pad_leading(Integer.to_string(cp, 16), 4, "0")
  end
end
